File: java-client/src/test/java/com/couchbase/client/java/codec/RawJsonTranscoderTest.java
/*
 * Copyright (c) 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.java.codec;

import com.couchbase.client.core.error.DecodingFailureException;
import com.couchbase.client.core.msg.kv.CodecFlags;
import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

/**
 * Verifies the functionality of the {@link RawJsonTranscoder}.
 */
class RawJsonTranscoderTest {

  private static final Transcoder RAW_JSON_TRANSCODER = RawJsonTranscoder.INSTANCE;

  @Test
  void encodesJsonPassthroughString() {
    String input = "{\"hello\": true}";
    Transcoder.EncodedValue output = RAW_JSON_TRANSCODER.encode(input);
    assertArrayEquals(input.getBytes(StandardCharsets.UTF_8), output.encoded());
  }

  @Test
  void encodesJsonPassthroughByteArray() {
    byte[] input = "{\"hello\": true}".getBytes(StandardCharsets.UTF_8);
    Transcoder.EncodedValue output = RAW_JSON_TRANSCODER.encode(input);
    assertArrayEquals(input, output.encoded());
  }

  @Test
  void decodesJsonPassthroughString() {
    String input = "{\"hello\": true}";
    String output = RAW_JSON_TRANSCODER.decode(
      String.class,
      input.getBytes(StandardCharsets.UTF_8),
      CodecFlags.JSON_COMPAT_FLAGS
    );
    assertEquals(input, output);
  }

  @Test
  void rejectsDecodingAsObject() {
    assertThrows(DecodingFailureException.class, () ->
        RAW_JSON_TRANSCODER.decode(
            Object.class,
            "{\"hello\": true}".getBytes(StandardCharsets.UTF_8),
            CodecFlags.JSON_COMPAT_FLAGS));
  }

  @Test
  void decodesJsonPassthroughByteArray() {
    byte[] input = "{\"hello\": true}".getBytes(StandardCharsets.UTF_8);
    byte[] output = RAW_JSON_TRANSCODER.decode(
      byte[].class,
      input,
      CodecFlags.JSON_COMPAT_FLAGS
    );
    assertArrayEquals(input, output);
  }

}

File: kotlin-client/src/test/kotlin/com/couchbase/client/kotlin/codec/RawJsonTranscoderTest.kt
/*
 * Copyright 2021 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.kotlin.codec

import com.couchbase.client.core.error.DecodingFailureException
import com.couchbase.client.core.error.InvalidArgumentException
import com.couchbase.client.core.msg.kv.CodecFlags
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

internal class RawJsonTranscoderTest {

    private val transcoder: Transcoder = RawJsonTranscoder

    @Test
    fun `echoes pre-encoded content`() {
        val content = Content.serializedJavaObject("xyzzy".toByteArray())
        assertSame(content, transcoder.encode(content))
    }

    @Test
    fun `fails to encode null`() {
        assertThrows<InvalidArgumentException> {
            transcoder.encode<String?>(null)
        }
    }

    @Test
    fun `can encode string`() {
        val serialized = transcoder.encode("xyzzy")
        assertArrayEquals("xyzzy".toByteArray(), serialized.bytes)
        assertEquals(CodecFlags.JSON_COMPAT_FLAGS, serialized.flags)
    }

    @Test
    fun `can decode as string`() {
        assertEquals("xyzzy", transcoder.decode<String>(Content.json("xyzzy")))
    }

    @Test
    fun `can encode bytes`() {
        val serialized = transcoder.encode("xyzzy".toByteArray())
        assertArrayEquals("xyzzy".toByteArray(), serialized.bytes)
        assertEquals(CodecFlags.JSON_COMPAT_FLAGS, serialized.flags)
    }

    @Test
    fun `can decode as bytes`() {
        assertArrayEquals("xyzzy".toByteArray(), transcoder.decode<ByteArray>(Content.json("xyzzy")))
    }

    @Test
    fun `fails to encode list`() {
        assertThrows<InvalidArgumentException> {
            transcoder.encode(Object())
        }
    }

    @Test
    fun `fails to decode as list`() {
        assertThrows<DecodingFailureException> {
            transcoder.decode<List<String>>(Content.json("[]"))
        }
    }
}


